go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -434,6 +434,7 @@ func (vr *vreplicator) readSettings(ctx context.Context, dbClient *vdbClient) (s
 }
 
 func (vr *vreplicator) setMessage(message string) error {
+	message = strings.ToValidUTF8(message, "�")
```

Comment:
I don't understand this line.

Reply:
`strings.ToValidUTF8` replaces all non-UTF8 characters in `message` with `�`, which is the same character you see on your terminal or in CI logs etc. for non-UTF8 characters. The main thing is to remove non-UTF8 characters, of course.

Reply:
I thought that strings in go were already utf8 encoded and that unicode character is used for invalid characters. Do you have any examples of what we're trying to fix?

Reply:
Yeah, so the story is this: we've seen a production issue where a `vitess` Online DDL failed, and the error message was `Picked source tablet...`, which is really not an error message. Upon investigation, we found a log entry saying (details redacted):

```
E0712 14:51:18.837571       1 controller.go:268] vreplication stream 564 going into error state due to Column 'c3' cannot be null (errno 1048) (sqlstate 23000) during query: insert into _4e3e77b5_d1bc_5366_86d2_cdb474f41540_20230712145011_vrepl(c1, c2, c3, c4, c5)
```

Now, the underlying problem was that the migration converted a column from `NULL`able to `NOT NULL`. So far so good. The table data did in fact have `NULL` data. OK fine, too bad, but we do handle this scenario and know how to fail the migration. Which we did. 

But then the problem was that calling `setState` (updating `_vt.vreplication.message`) failed. So we were unable to store the actual error message in Vitess's DB.

Why would that happen? I was reminded of this: https://github.com/vitessio/vitess/issues/13487 ; when you attempt to write an arbitrary binary, that can fail. The error message contains specific table data, which could be binary. This is why in https://github.com/vitessio/vitess/issues/13487 we converted non UTF8 characters to `�`. I am unable to determine the precise data that caused this failure, but `13487` seems very relevant.

Moreover, Vitess did not print out the `setState` (i.e. the `UPDATE _vt.vreplication`) error. This PR therefore:

1. Ensures to log exactly what we were trying to write, and what precise error we got back.
2. Attempt to sanitize the input in the first place.

Reply:
OK, it doesn't hurt to do this. I would note that the field in the table is varbinary too. That may be a relevant factor here.